api/v1alpha1/akodeploymentconfig_webhook.go

```diff
@@ -34,64 +46,267 @@ func (r *AKODeploymentConfig) ValidateCreate() error {
 }
 
-// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
-func (r *AKODeploymentConfig) ValidateUpdate(old runtime.Object) error {
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *AKODeploymentConfig) ValidateDelete() error {
+	akoDeploymentConfigLog.Info("validate delete", "name", r.Name)
+	return nil
+}
+
+// validateCluserSelector checks AKODeploymentConfig object's cluster selector field input is valid or not
+// when old is nil, it is used for AKODeploymentConfig object create, otherwise it is used for AKODeploymentConfig
+// object update
+func (r *AKODeploymentConfig) validateCluserSelector(old *AKODeploymentConfig) field.ErrorList {
 	var allErrs field.ErrorList
+	// when update AKODeploymentConfig object, cluster selector should be immutable
+	if old != nil {
+		if old.Spec.ClusterSelector.String() != r.Spec.ClusterSelector.String() {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "ClusterSelector"),
+				r.Spec.ClusterSelector,
+				"field should not be changed"))
+			return allErrs
+		}
+	}
+	// convert cluster selector to label selector
+	selector, err := metav1.LabelSelectorAsSelector(&r.Spec.ClusterSelector)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "ClusterSelector"),
+			r.Spec.ClusterSelector,
+			err.Error()))
+	}
+	// non default ADC (a.k.a name is not install-ako-for-all), should have non-empty cluster selector
+	if r.ObjectMeta.Name != WorkloadClusterAkoDeploymentConfig && selector.Empty() {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "ClusterSelector"),
+			r.Spec.ClusterSelector,
+			"field should not be empty for non-default ADC"))
+	}
+	return allErrs
+}
 
-	akoDeploymentConfigLog.Info("validate update", "name", r.Name)
+// validateAVI checks all NSX Advanced Load Balancer related fields are valid or not
+// when old is nil, it is used for AKODeploymentConfig object create, otherwise it is used for AKODeploymentConfig
+// object update. Following fields are already required fileds in CRD, so no need to check if those fields are empty.
+// - adminCredentialRef
+// - certificateAuthorityRef
+// - cloudName
+// - controller
+// - dataNetwork
+// - serviceEngineGroup
+func (r *AKODeploymentConfig) validateAVI(old *AKODeploymentConfig) field.ErrorList {
+	var allErrs field.ErrorList
 
-	oldADC, ok := old.(*AKODeploymentConfig)
+	// check avi related secret
+	adminCredential := &corev1.Secret{}
+	if err := r.validateAviSecret(adminCredential, r.Spec.AdminCredentialRef); err != nil {
+		allErrs = append(allErrs, err)
+	}
 
-	if !ok {
-		return apierrors.NewBadRequest(fmt.Sprintf("expected a AKODeploymentConfig but got a %T", old))
+	aviControllerCA := &corev1.Secret{}
+	if err := r.validateAviSecret(aviControllerCA, r.Spec.CertificateAuthorityRef); err != nil {
```

Comment:
suggest: i've seen user mistakenly use private key as the CA. maybe check wether the secret contains a valid certificate. at least the format should be correct

Reply:
Same for this one : https://github.com/vmware-tanzu/load-balancer-operator-for-kubernetes/pull/94#discussion_r907008169

Reply:
Give a second thought on this, we already verified if users can use the certificate they provide to connect to avi controller, so if it is not a valid certificate, it will throw error.